{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4401a10-080d-44d4-a85e-3ae9dfd2d7e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "from matplotlib import cm\n",
    "import matplotlib.patches as mpatches\n",
    "import json\n",
    "import h5py\n",
    "\n",
    "from itertools import combinations\n",
    "\n",
    "\n",
    "bnsi_path = '/todo_your_path_to_bonsai/Bonsai-data-representation'\n",
    "sys.path.append(bnsi_path)\n",
    "from bonsai_scout.bonsai_scout_helpers import Bonvis_figure, Bonvis_settings, Bonvis_metadata, Annotation_info, get_celltype_colors_new\n",
    "\n",
    "def cbar_to_numb(val):\n",
    "    if cbar_info['log']:\n",
    "        return np.exp(val * (max_val - min_val) + min_val)\n",
    "    return val * (max_val - min_val) + min_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1db936c-babb-4f19-936b-f7a81a625b1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99521b64-5ebc-437b-a000-6d999170b4cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_output = \"todo_path_to_dataset\"\n",
    "\n",
    "sanity_out_path = os.path.join(path_to_output, \"Sanity\")\n",
    "bonsai_results_folders = os.path.join(path_to_output, \"bonsai\")\n",
    "save_dir = \"./figures\"\n",
    "if not os.path.exists(save_dir):\n",
    "    os.mkdir(save_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b94365e-aa92-4cb2-8dcf-7cc6b02660cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## bonsai posterior estiamtes of ltqs\n",
    "posterior_est_var_file = \"posterior_ltqsVars_vertByGene.npy\"  \n",
    "posterior_est_file = \"posterior_ltqs_vertByGene.npy\"\n",
    "\n",
    "bonsai_dir = os.path.join(path_to_output, \"bonsai\")\n",
    "\n",
    "# GET POSTERIOR ESTIMATES FOR ALL GENES\n",
    "results_dir = \"todo_path_to_posterior_files\"\n",
    "\n",
    "cellid_file = os.path.join(path_to_output, \"Sanity\", \"cellID.txt\")\n",
    "cell_id_true_df = pd.read_csv(cellid_file, header=None, names=[\"cell_id\"])\n",
    "cell_id_true_df[\"cell_id_idx\"] = cell_id_true_df.index\n",
    "\n",
    "ltq_post_est = np.load(os.path.join(bonsai_dir, results_dir, posterior_est_file))\n",
    "ltq_eb_post_est = np.load(os.path.join(bonsai_dir, results_dir, posterior_est_var_file))\n",
    "\n",
    "with open(os.path.join(bonsai_dir, results_dir,\"metadata.json\")) as f:\n",
    "    metadata = json.load(f)\n",
    "\n",
    "vert_info = pd.read_csv(os.path.join(bonsai_dir, results_dir, \"vertInfo.txt\"), sep=\"\\t\")\n",
    "\n",
    "vert_info_sub = pd.merge(vert_info, cell_id_true_df, left_on=\"vertName\", right_on=\"cell_id\")\n",
    "vert_info_sub = vert_info_sub.sort_values(\"cell_id_idx\")\n",
    "\n",
    "ltq_post_est_sub = ltq_post_est[vert_info_sub.vertInd, :]\n",
    "ltq_post_est_sub = ltq_post_est_sub.T\n",
    "print(ltq_post_est_sub.shape)\n",
    "\n",
    "ltq_post_est_sub_df = pd.DataFrame(ltq_post_est_sub, index = metadata[\"geneIds\"], columns=vert_info_sub.vertName)\n",
    "ltq_post_est_sub_df = ltq_post_est_sub_df.T\n",
    "\n",
    "\n",
    "\n",
    "ltq_post_est_sub_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf08e3df-4e8d-4d45-8af0-bb9d6010a079",
   "metadata": {},
   "source": [
    "## Sum up counts per transcript and define NK thresholds\n",
    "Marker genes: \"CMC1\",\n",
    "         \"CST7\",\n",
    "         \"GNLY\",\n",
    "         \"KLRF1\",\n",
    "         \"NKG7\",\n",
    "         \"PRF1\",\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da564331-bc55-48cd-a6b4-f08e8ff06493",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"NKG7_trscrpt1\", \"NKG7_trscrpt2\", \n",
    "         \"GNLY_trscrpt1\", \"GNLY_trscrpt2\", \"GNLY_trscrpt3\",\n",
    "         \"CMC1\", \"CST7\", \"KLRF1_trscrpt1\",\n",
    "         \"PRF1\"\n",
    "        ]\n",
    "\n",
    "## sum transcripts\n",
    "expr = ltq_post_est_sub_df[genes]\n",
    "\n",
    "expr_exp = np.exp(expr)\n",
    "\n",
    "# remove transcript suffixes (e.g. \"_trscrpt1\", \"_trscrpt2\")\n",
    "expr_exp.columns = expr_exp.columns.str.replace(r\"_trscrpt\\d+\", \"\", regex=True)\n",
    "\n",
    "# group by gene name (columns with same gene) and sum\n",
    "expr_sum = expr_exp.groupby(expr_exp.columns, axis=1).sum()\n",
    "\n",
    "df_processed = np.log(expr_sum)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c0d8a96-8c55-4e3d-a56c-723f3f449ff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67c75bf1-e7fa-4a7d-b484-ac05503a9923",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, ax = plt.subplots(1,6, figsize=(12,2), dpi=200)\n",
    "\n",
    "# decided thresholds\n",
    "NKG7_threshold = -7.5\n",
    "GNLY_threshold = np.percentile(df_processed[\"GNLY\"], 90)\n",
    "PRF1_threshold = -9.2 \n",
    "CMC1_threshold = -8.5\n",
    "CST7_threshold = -8.7\n",
    "KLRF1_threshold = np.percentile(df_processed[\"KLRF1\"], 90)\n",
    "\n",
    "log=True\n",
    "\n",
    "thresholds = {\"CMC1\": CMC1_threshold,\n",
    "              \"CST7\": CST7_threshold,\n",
    "         \"GNLY\": GNLY_threshold,\n",
    "         \"KLRF1\": KLRF1_threshold,\n",
    "         \"NKG7\": NKG7_threshold,\n",
    "         \"PRF1\": PRF1_threshold,\n",
    "              \n",
    "             }\n",
    "genes = [\"NKG7\",\n",
    "         \"GNLY\",\n",
    "        \"KLRF1\",\n",
    "         \"PRF1\",\n",
    "         \"CMC1\",\n",
    "         \"CST7\",\n",
    "        ]\n",
    "\n",
    "\n",
    "ax = ax.flatten()\n",
    "for ax_idx, gene in enumerate(genes):\n",
    "\n",
    "    ax[ax_idx].hist(df_processed[gene], bins=100, log=log, density=True, alpha=0.5, color=\"grey\")\n",
    "    ax[ax_idx].axvline(thresholds[gene], color=\"black\", linewidth=1)\n",
    "    ax[ax_idx].text(0.5, 1.20, gene, fontsize=14, ha='center', transform=ax[ax_idx].transAxes)\n",
    "    ax[ax_idx].text(0.5, 1.04, \"threshold= {}\".format(np.round(thresholds[gene], 5)), fontsize=10, ha='center', transform=ax[ax_idx].transAxes)\n",
    "\n",
    "\n",
    "fig.text(0.5, 0.0, \"log transcription quotients\", ha='center', va='center')\n",
    "\n",
    "fig.text(0, 0.5, 'density', ha='center', va='center', rotation='vertical')\n",
    "   \n",
    "plt.tight_layout()\n",
    "plt.savefig('{}/marker_gene_expr-all_cells-histogram-stepfilled-log_{}.svg'.format(save_dir, log), bbox_inches='tight')\n",
    "plt.savefig('{}/marker_gene_expr-all_cells-histogram-stepfilled-log_{}.png'.format(save_dir, log), bbox_inches='tight')\n",
    "    \n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e498c68-07d6-49a2-b23a-afc808136075",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"NKG7_threshold: {}\".format(NKG7_threshold))\n",
    "print(\"GNLY_threshold: {}\".format(GNLY_threshold))\n",
    "print(\"PRF1_threshold: {}\".format(PRF1_threshold))\n",
    "print(\"CMC1_threshold: {}\".format(CMC1_threshold))\n",
    "print(\"CST7_threshold: {}\".format(CST7_threshold))\n",
    "print(\"KLRF1_threshold: {}\".format(KLRF1_threshold))\n",
    "\n",
    "\n",
    "ltq_post_est_sub_df[\"NK_nkg7_gnly_prf1_cmc1_cst7_klrf1\"] = np.where((df_processed[\"NKG7\"] > NKG7_threshold) & \n",
    "                                                                    (df_processed[\"GNLY\"] > GNLY_threshold) & \n",
    "                                                                    (df_processed[\"PRF1\"] > PRF1_threshold) &\n",
    "                                                                    (df_processed[\"CMC1\"] > CMC1_threshold) &\n",
    "                                                                    (df_processed[\"CST7\"] > CST7_threshold) &\n",
    "                                                                    (df_processed[\"KLRF1\"] > KLRF1_threshold) \n",
    "                                                                    , \"NK\", \"other\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7af1c72-ac98-4219-9ce1-35ddb1eba2a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "NK_def = \"NK_nkg7_gnly_prf1_cmc1_cst7_klrf1\"\n",
    "tmp = ltq_post_est_sub_df[ltq_post_est_sub_df[NK_def] == \"NK\"]\n",
    "x = len(tmp)\n",
    "print(\"{}: num of NK cells: {}\".format(NK_def, x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed52858a-1c03-4b6c-b638-9cd2cb1db202",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5fdacf1a-a2b7-4317-8c2d-b269f28a1ad5",
   "metadata": {},
   "source": [
    "## Load bonsai tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d7c4d82-e37f-4a4b-bf1b-c9c413184edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%capture  \n",
    "print(bonsai_results_folders)\n",
    "data_path = os.path.join(bonsai_results_folders, 'bonsai_vis_data.hdf')\n",
    "settings_path = os.path.join(bonsai_results_folders, 'bonsai_vis_settings.json')\n",
    "\n",
    "print(\"read in file: {}\".format(data_path))\n",
    "bonvis_metadata = Bonvis_metadata(data_path)\n",
    "bonvis_settings = Bonvis_settings(load_settings_path=settings_path)\n",
    "bonvis_data_hdf = h5py.File(data_path, 'r')\n",
    "\n",
    "# Set darker edges\n",
    "bonvis_settings.edge_style = {\"color\": [0.4117647058823529, 0.4117647058823529, 0.4117647058823529, 1.0],\n",
    "                             'linewidth':0.6}\n",
    "\n",
    "\n",
    "bonvis_fig = Bonvis_figure(bonvis_data_hdf, \n",
    "                           bonvis_metadata, \n",
    "                           bonvis_data_path=data_path,\n",
    "                           bonvis_settings=bonvis_settings)\n",
    "celltype_info = bonvis_fig.bonvis_settings.celltype_info\n",
    "\n",
    "\n",
    "\n",
    "bonvis_fig.create_figure(figsize=(6, 6))\n",
    "\n",
    "# Here, we set the desired celltype-annotation for the dataset\n",
    "node_style = celltype_info.annot_infos['annot_{}'.format(NK_def)].label\n",
    "\n",
    "# Visualize the tree in the ly_dendrogram layout, with the correct celltype-annotation\n",
    "geometry = 'flat'\n",
    "ly_type = \"ly_dendrogram_ladderized\"\n",
    "\n",
    "\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=node_style, \n",
    "                         # only show NK cells\n",
    "                         renew_mask=False,\n",
    "                          scale_nodes=0.6666666666666666,\n",
    "                         tweak_inds=1911, # change root # at node internal_1095\n",
    "                         new_flip_id=True\n",
    "                        )\n",
    "bonvis_fig.set_mask_for_subset([{'type': 'annot', 'info': 'NK', 'mask_is_on': True}])\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=node_style, \n",
    "                         # only show NK cells\n",
    "                         renew_mask=True,\n",
    "                          # scale_nodes=0.6666666666666666,\n",
    "                         tweak_inds=1911, # change root # at node internal_1095\n",
    "                         new_flip_id=True\n",
    "                        )\n",
    "\n",
    "bonsai_fig_darker_lines = bonvis_fig.create_figure(figsize=(6, 6))\n",
    "\n",
    "bonsai_fig_darker_lines.savefig(\"{}/bonsai_dendogram-NK_annot-darker_lines_full.svg\".format(save_dir))\n",
    "bonsai_fig_darker_lines.savefig(\"{}/bonsai_dendogram-NK_annot-darker_lines_full.png\".format(save_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc1fb81c-f78c-42f7-afac-8fbd8b23ac52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create new bonvisfig\n",
    "gene = \"LYZ\" # high expressed in myeloid\n",
    "print(gene)\n",
    "bonvis_fig.bonvis_settings.node_style['feature_path'] = \"data/normalized\"\n",
    "bonvis_settings.node_style['use_mask'] = False\n",
    "geometry = 'flat'\n",
    "ly_type = \"ly_dendrogram_ladderized\"\n",
    "\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=gene, \n",
    "                         renew_mask=True,\n",
    "                         tweak_inds=1911,\n",
    "                        )\n",
    "\n",
    "outfig = bonvis_fig.create_figure(figsize=(6, 6))\n",
    "\n",
    "outfig.savefig(\"{}/bonsai_dendogram-gene_{}-darker_lines_full.svg\".format(save_dir, gene))\n",
    "outfig.savefig(\"{}/bonsai_dendogram-gene_{}-darker_lines_full.png\".format(save_dir, gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0351bb98-f7eb-4bbb-97fa-3172b79581c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(gene)\n",
    "annot_info = bonvis_settings.node_style['annot_info']\n",
    "cbar_info = annot_info.cbar_info\n",
    "cbar_info\n",
    "\n",
    "\n",
    "fig_cbar = plt.figure(figsize=(2, 4))\n",
    "ax_cbar = fig_cbar.add_subplot(111)\n",
    "\n",
    "norm = colors.Normalize(vmin=0, vmax=1)\n",
    "mappable = cm.ScalarMappable(norm=norm, cmap=cbar_info['cmap'])\n",
    "cbar = plt.colorbar(mappable, cax=ax_cbar, orientation='vertical')\n",
    "\n",
    "\n",
    "print(cbar.ax.get_yticks())\n",
    "tick_list = [mappable.colorbar.vmin + t * (mappable.colorbar.vmax - mappable.colorbar.vmin) for t in [0,0.25, 0.5,0.75,1]]\n",
    "\n",
    "min_val = cbar_info['vmin']\n",
    "max_val = cbar_info['vmax']\n",
    "\n",
    "tick_labels = ['{:.2e}'.format(cbar_to_numb(tick)) for tick in tick_list]\n",
    "tick_labels = ['{:.2}'.format(cbar_to_numb(tick)) for tick in tick_list]\n",
    "cbar.set_ticks(tick_list)\n",
    "cbar.set_ticklabels(tick_labels)\n",
    "cbar.set_label('gene expression', fontsize=12, labelpad=10)\n",
    "plt.title(gene)\n",
    "plt.tight_layout()\n",
    "\n",
    "print('saving {}'.format(\"{}/colorbar-gene_{}.svg\".format(save_dir, gene)))\n",
    "fig_cbar.savefig(\"{}/colorbar-gene_{}.svg\".format(save_dir, gene))\n",
    "fig_cbar.savefig(\"{}/colorbar-gene_{}.png\".format(save_dir, gene))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2af0111e-1d9a-48eb-98ca-93e465c18900",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create new bonvisfig\n",
    "gene = \"IL7R_trscrpt2\" # high expressed in lymphoid\n",
    "print(gene)\n",
    "bonvis_fig.bonvis_settings.node_style['feature_path'] = \"data/normalized\"\n",
    "bonvis_settings.node_style['use_mask'] = False\n",
    "geometry = 'flat'\n",
    "ly_type = \"ly_dendrogram_ladderized\"\n",
    "\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=gene, \n",
    "                         renew_mask=True,\n",
    "                         tweak_inds=1911,\n",
    "                        )\n",
    "\n",
    "outfig = bonvis_fig.create_figure(figsize=(6, 6))\n",
    "\n",
    "outfig.savefig(\"{}/bonsai_dendogram-gene_{}-darker_lines_full.svg\".format(save_dir, gene))\n",
    "outfig.savefig(\"{}/bonsai_dendogram-gene_{}-darker_lines_full.png\".format(save_dir, gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f455bf59-6e12-4ba3-952d-26aa59fabbac",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(gene)\n",
    "annot_info = bonvis_settings.node_style['annot_info']\n",
    "cbar_info = annot_info.cbar_info\n",
    "cbar_info\n",
    "\n",
    "\n",
    "fig_cbar = plt.figure(figsize=(2, 4))\n",
    "ax_cbar = fig_cbar.add_subplot(111)\n",
    "\n",
    "norm = colors.Normalize(vmin=0, vmax=1)\n",
    "mappable = cm.ScalarMappable(norm=norm, cmap=cbar_info['cmap'])\n",
    "cbar = plt.colorbar(mappable, cax=ax_cbar, orientation='vertical')\n",
    "\n",
    "\n",
    "print(cbar.ax.get_yticks())\n",
    "tick_list = [mappable.colorbar.vmin + t * (mappable.colorbar.vmax - mappable.colorbar.vmin) for t in [0,0.25, 0.5,0.75,1]]\n",
    "\n",
    "min_val = cbar_info['vmin']\n",
    "max_val = cbar_info['vmax']\n",
    "\n",
    "tick_labels = ['{:.2e}'.format(cbar_to_numb(tick)) for tick in tick_list]\n",
    "tick_labels = ['{:.2}'.format(cbar_to_numb(tick)) for tick in tick_list]\n",
    "cbar.set_ticks(tick_list)\n",
    "cbar.set_ticklabels(tick_labels)\n",
    "cbar.set_label('gene expression', fontsize=12, labelpad=10)\n",
    "plt.title(gene)\n",
    "plt.tight_layout()\n",
    "\n",
    "print('saving {}'.format(\"{}/colorbar-gene_{}.svg\".format(save_dir, gene)))\n",
    "fig_cbar.savefig(\"{}/colorbar-gene_{}.svg\".format(save_dir, gene))\n",
    "fig_cbar.savefig(\"{}/colorbar-gene_{}.png\".format(save_dir, gene))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ef26689-206e-4ec3-91ef-8a5fdfc52f09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create new bonvisfig\n",
    "gene = \"CD11c\"\n",
    "print(gene)\n",
    "bonvis_fig.bonvis_settings.node_style['feature_path'] = \"data/at_cnts_CLR_transformed\"\n",
    "bonvis_settings.node_style['use_mask'] = False\n",
    "geometry = 'flat'\n",
    "ly_type = \"ly_dendrogram_ladderized\"\n",
    "\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=gene, \n",
    "                         renew_mask=True,\n",
    "                         tweak_inds=1911,\n",
    "                        )\n",
    "\n",
    "outfig = bonvis_fig.create_figure(figsize=(6, 6))\n",
    "\n",
    "outfig.savefig(\"{}/bonsai_dendogram-antibody_{}-darker_lines_full.svg\".format(save_dir, gene))\n",
    "outfig.savefig(\"{}/bonsai_dendogram-antibody_{}-darker_lines_full.png\".format(save_dir, gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e098f1fe-791c-49c9-8336-96f2af068ffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(gene)\n",
    "annot_info = bonvis_settings.node_style['annot_info']\n",
    "cbar_info = annot_info.cbar_info\n",
    "cbar_info\n",
    "\n",
    "\n",
    "fig_cbar = plt.figure(figsize=(2, 4))\n",
    "ax_cbar = fig_cbar.add_subplot(111)\n",
    "\n",
    "norm = colors.Normalize(vmin=0, vmax=1)\n",
    "mappable = cm.ScalarMappable(norm=norm, cmap=cbar_info['cmap'])\n",
    "cbar = plt.colorbar(mappable, cax=ax_cbar, orientation='vertical')\n",
    "\n",
    "\n",
    "print(cbar.ax.get_yticks())\n",
    "tick_list = [mappable.colorbar.vmin + t * (mappable.colorbar.vmax - mappable.colorbar.vmin) for t in [0,0.25, 0.5,0.75,1]]\n",
    "\n",
    "min_val = cbar_info['vmin']\n",
    "max_val = cbar_info['vmax']\n",
    "\n",
    "tick_labels = ['{:.2e}'.format(cbar_to_numb(tick)) for tick in tick_list]\n",
    "tick_labels = ['{:.2}'.format(cbar_to_numb(tick)) for tick in tick_list]\n",
    "cbar.set_ticks(tick_list)\n",
    "cbar.set_ticklabels(tick_labels)\n",
    "cbar.set_label('normalized surface protein expression', fontsize=12, labelpad=10)\n",
    "plt.title(gene)\n",
    "plt.tight_layout()\n",
    "\n",
    "print('saving {}'.format(\"{}/colorbar-antibody_{}.svg\".format(save_dir, gene)))\n",
    "fig_cbar.savefig(\"{}/colorbar-antibody_{}.svg\".format(save_dir, gene))\n",
    "fig_cbar.savefig(\"{}/colorbar-antibody_{}.png\".format(save_dir, gene))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d8635aa-81a6-46ca-a041-b7d8fe3775c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create new bonvisfig\n",
    "gene = \"CD3\"\n",
    "print(gene)\n",
    "bonvis_fig.bonvis_settings.node_style['feature_path'] = \"data/at_cnts_CLR_transformed\"\n",
    "bonvis_settings.node_style['use_mask'] = False\n",
    "geometry = 'flat'\n",
    "ly_type = \"ly_dendrogram_ladderized\"\n",
    "\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=gene, \n",
    "                         renew_mask=True,\n",
    "                         tweak_inds=1911,\n",
    "                        )\n",
    "\n",
    "outfig = bonvis_fig.create_figure(figsize=(6, 6))\n",
    "\n",
    "outfig.savefig(\"{}/bonsai_dendogram-antibody_{}-darker_lines_full.svg\".format(save_dir, gene))\n",
    "outfig.savefig(\"{}/bonsai_dendogram-antibody_{}-darker_lines_full.png\".format(save_dir, gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1ea5846-7321-409a-876c-f5e1422cbd7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(gene)\n",
    "annot_info = bonvis_settings.node_style['annot_info']\n",
    "cbar_info = annot_info.cbar_info\n",
    "cbar_info\n",
    "\n",
    "\n",
    "fig_cbar = plt.figure(figsize=(2, 4))\n",
    "ax_cbar = fig_cbar.add_subplot(111)\n",
    "\n",
    "norm = colors.Normalize(vmin=0, vmax=1)\n",
    "mappable = cm.ScalarMappable(norm=norm, cmap=cbar_info['cmap'])\n",
    "cbar = plt.colorbar(mappable, cax=ax_cbar, orientation='vertical')\n",
    "\n",
    "\n",
    "print(cbar.ax.get_yticks())\n",
    "tick_list = [mappable.colorbar.vmin + t * (mappable.colorbar.vmax - mappable.colorbar.vmin) for t in [0,0.25, 0.5,0.75,1]]\n",
    "\n",
    "min_val = cbar_info['vmin']\n",
    "max_val = cbar_info['vmax']\n",
    "\n",
    "tick_labels = ['{:.2e}'.format(cbar_to_numb(tick)) for tick in tick_list]\n",
    "tick_labels = ['{:.2}'.format(cbar_to_numb(tick)) for tick in tick_list]\n",
    "cbar.set_ticks(tick_list)\n",
    "cbar.set_ticklabels(tick_labels)\n",
    "cbar.set_label('normalized surface protein expression', fontsize=12, labelpad=10)\n",
    "plt.title(gene)\n",
    "plt.tight_layout()\n",
    "\n",
    "print('saving {}'.format(\"{}/colorbar-antibody_{}.svg\".format(save_dir, gene)))\n",
    "fig_cbar.savefig(\"{}/colorbar-antibody_{}.svg\".format(save_dir, gene))\n",
    "fig_cbar.savefig(\"{}/colorbar-antibody_{}.png\".format(save_dir, gene))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1355112e-d693-41c1-8c69-b4c26b4fc874",
   "metadata": {},
   "source": [
    "#### myeloid and lymphoid clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60eaa4a4-5075-48fd-967b-1aec5fb64335",
   "metadata": {},
   "outputs": [],
   "source": [
    "# myleoid clade\n",
    "curr_subset = {}\n",
    "curr_subset['type'] = \"subtree\"\n",
    "curr_subset['info'] = (9008, 14099) \n",
    "nk_cell_inds_group1 = bonvis_fig.get_cell_inds_in_subset(curr_subset)[0]\n",
    "nk_cell_ids_group1 = np.array(bonvis_metadata.cell_ids)[nk_cell_inds_group1]\n",
    "\n",
    "curr_subset['mask_is_on']= True\n",
    "bonvis_fig.set_mask_for_subset([curr_subset])\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=node_style, \n",
    "                         # only show NK cells\n",
    "                         renew_mask=True,\n",
    "                          # scale_nodes=0.6666666666666666,\n",
    "                         tweak_inds=1911, # change root\n",
    "                        )\n",
    "bonsai_fig_darker_lines = bonvis_fig.create_figure(figsize=(6, 6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e837127b-b568-4a4d-881d-d2eacce8d053",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lymphoid clade\n",
    "curr_subset = {}\n",
    "curr_subset['type'] = \"subtree\"\n",
    "curr_subset['info'] = (17988, 26116) \n",
    "nk_cell_inds_group2 = bonvis_fig.get_cell_inds_in_subset(curr_subset)[0]\n",
    "nk_cell_ids_group2 = np.array(bonvis_metadata.cell_ids)[nk_cell_inds_group2]\n",
    "\n",
    "curr_subset['mask_is_on']= True\n",
    "bonvis_fig.set_mask_for_subset([curr_subset])\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=node_style, \n",
    "                         # only show NK cells\n",
    "                         renew_mask=True,\n",
    "                         tweak_inds=1911, # change root\n",
    "                        )\n",
    "bonsai_fig_darker_lines = bonvis_fig.create_figure(figsize=(6, 6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a1e217d-c97f-4e16-a9a7-d71988908eb3",
   "metadata": {},
   "source": [
    "#### Definition of myeloid and lymphoid NK cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c89bf80-79a2-4457-94f5-e3a4b0355af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_celltype = \"NK\"\n",
    "group1 = \"NK_myeloid\"\n",
    "group2 = \"NK_lymphoid\"\n",
    "group3 = \"other\"\n",
    "\n",
    "group = \"NK_group\"\n",
    "ltq_post_est_sub_df[group] = [ group3 if ct != my_celltype\n",
    "               else group1 if cb in nk_cell_ids_group1 \n",
    "               else group2 if cb in nk_cell_ids_group2 \n",
    "               else group3 \n",
    "               for cb, ct in zip(ltq_post_est_sub_df.index, ltq_post_est_sub_df[NK_def])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15319697-ff6e-42c4-a375-d11c965047fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "ltq_post_est_sub_df.groupby(group).count()[NK_def]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11ee61fa-403a-46a2-b974-cd0ed84f7e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "#red\n",
    "color_NK_g1_html = \"#FF0000\"\n",
    "color_NK_g1_rgba = (1.0, 0.0, 0.0, 0.8)\n",
    "\n",
    "#blue\n",
    "color_NK_g2_html = \"#0000FF\"\n",
    "color_NK_g2_rgba = (0.0, 0.0, 1.0, 0.8)\n",
    "\n",
    "color_other_html = '#808080'\n",
    "color_other_rgba = (0.5019607843137255, 0.5019607843137255, 0.5019607843137255, 0.5)\n",
    "\n",
    "colors_rgba = {\"NK_lymphoid\": color_NK_g1_rgba, \"NK_myeloid\": color_NK_g2_rgba, \"other\":color_other_rgba}\n",
    "NK_colors = {\"NK_lymphoid\": \"red\", \"NK_myeloid\": \"blue\", \"other\":\"grey\"}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82badc7c-edef-484a-9064-bd1c3fe0878c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cats = list(ltq_post_est_sub_df[group].unique())\n",
    "print(cats)\n",
    "annot_to_color = {'NK_lymphoid': colors_rgba[\"NK_lymphoid\"], \n",
    "                  'NK_myeloid':colors_rgba[\"NK_myeloid\"],\n",
    "                  'other': (0.8274509803921568, 0.8274509803921568, 0.8274509803921568, 0.1)} #lightgray\n",
    "\n",
    "\n",
    "cbar_info = {'cmap': None, 'vmin': None, 'vmax': None, 'log': None}\n",
    "label = \"NK_group\"\n",
    "info_key = \"annot_NK_group\"\n",
    "annot_type = 'cellstates' \n",
    "info_object = 'cs_info_dict'\n",
    "\n",
    "cl_annot = Annotation_info(cats=cats, annot_to_color=annot_to_color, label=label,\n",
    "                                   cbar_info=cbar_info, annot_type=annot_type,\n",
    "                                   color_type='categorical', info_object=info_object,\n",
    "                                   info_key=info_key)\n",
    "cl_annot.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaf14b2e-0256-4b97-baca-8f4549e462c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "bonvis_metadata.cs_info['cs_info_dict'][info_key] = ltq_post_est_sub_df[group].tolist()\n",
    "\n",
    "bonvis_settings.set_annot(annot_info=cl_annot)\n",
    "bonvis_settings.cell_to_celltype, _ = bonvis_fig.get_color_info(annot_info=bonvis_fig.bonvis_settings.node_style['annot_info'])\n",
    "\n",
    "\n",
    "bonvis_settings.celltype_info.annot_infos[info_key] = cl_annot\n",
    "bonvis_settings.celltype_info.annot_alts.append(info_key)\n",
    "\n",
    "celltype_info = bonvis_fig.bonvis_settings.celltype_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1439801-c057-409e-9375-8b1356bfcf82",
   "metadata": {},
   "outputs": [],
   "source": [
    "myannot = \"annot_NK_group\"\n",
    "sub1='NK_lymphoid'\n",
    "sub2='NK_myeloid'\n",
    "\n",
    "sub_idx1 = np.argwhere(np.array(celltype_info.annot_infos[myannot].cats) == sub1)[0][0]\n",
    "sub_idx2 = np.argwhere(np.array(celltype_info.annot_infos[myannot].cats) == sub2)[0][0]\n",
    "\n",
    "print(sub_idx1)\n",
    "print(sub_idx2)\n",
    "\n",
    "selected_annot1 = celltype_info.annot_infos[myannot].cats[sub_idx1]\n",
    "selected_annot2 = celltype_info.annot_infos[myannot].cats[sub_idx2]\n",
    "print(selected_annot1)\n",
    "print(selected_annot2)\n",
    "\n",
    "selected_subset1 = {'type': 'annot', \n",
    "                       'info': selected_annot1, \n",
    "                       'mask_is_on': True}\n",
    "selected_subset2 = {'type': 'annot', \n",
    "                       'info': selected_annot2, \n",
    "                       'mask_is_on': True}\n",
    "\n",
    "selected_subset = [selected_subset1,selected_subset2 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b354f790-123c-4932-8663-83b34e5c75f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# With NK mask\n",
    "# Here, we set the desired celltype-annotation for the dataset\n",
    "node_style = celltype_info.annot_infos['annot_NK_group'].label\n",
    "\n",
    "geometry = 'flat'\n",
    "ly_type = \"ly_dendrogram_ladderized\"\n",
    "\n",
    "\n",
    "bonvis_fig.set_mask_for_subset(selected_subset)\n",
    "bonvis_fig.update_figure(ly_type=ly_type, \n",
    "                         geometry=geometry, \n",
    "                         node_style=node_style, \n",
    "                         # only show NK cells\n",
    "                         renew_mask=True,\n",
    "                         tweak_inds=1911,\n",
    "                        )\n",
    "\n",
    "nk_new_bonsai_fig_darker_lines = bonvis_fig.create_figure(figsize=(6, 6))\n",
    "nk_new_bonsai_fig_darker_lines.savefig(\"{}/bonsai_dendogram-nk_lymphoid_myeloid-darker_lines_full.svg\".format(save_dir))\n",
    "nk_new_bonsai_fig_darker_lines.savefig(\"{}/bonsai_dendogram-nk_lymphoid_myeloid-darker_lines_full.png\".format(save_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14a8a8b6-1cbf-4f92-a14e-526621dd501a",
   "metadata": {},
   "outputs": [],
   "source": [
    "categories = [\n",
    "    \"NK_lymphoid\",\n",
    "    \"NK_myleoid\",\n",
    "    \"rest\"\n",
    "]\n",
    "\n",
    "colors_nk_subgroups = [color_NK_g1_html, color_NK_g2_html, color_other_html]  # Use the first 13 colors\n",
    "\n",
    "\n",
    "legend_elements = [\n",
    "    Line2D([0], [0], marker='o', color='w', markerfacecolor=color, markersize=10, label=category, alpha=0.8)\n",
    "    for color, category in zip(colors_nk_subgroups, categories)\n",
    "]\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 3))  \n",
    "ax.axis('off') \n",
    "legend = ax.legend(handles=legend_elements, loc='center', frameon=False, ncol=1, fontsize=10)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cfc3645-2ae9-45bd-818b-55a70ecdb113",
   "metadata": {},
   "source": [
    "### Antibody expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f27b9c88-3076-49dd-b786-6ee2839e9517",
   "metadata": {},
   "outputs": [],
   "source": [
    "antibody_names = json.loads(bonvis_data_hdf[\"data/at_cnts_CLR_transformed\"].attrs['gene_ids'])\n",
    "cd56_idx = np.where(np.array(antibody_names) == \"CD56\")[0]\n",
    "cd16_idx = np.where(np.array(antibody_names) == \"CD16\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "927fac0c-d62e-4881-b0d8-af8b21166f9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "trnfs_antibodies = bonvis_data_hdf[\"data/at_cnts_CLR_transformed/means\"][:]\n",
    "\n",
    "s=15\n",
    "\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "grid = fig.add_gridspec(4, 4, hspace=0.1, wspace=0.1) \n",
    "\n",
    "\n",
    "main_ax = fig.add_subplot(grid[1:, :-1])\n",
    "\n",
    "\n",
    "main_ax.scatter(x=trnfs_antibodies[cd56_idx, ltq_post_est_sub_df.NK_group==group3],\n",
    "            y=trnfs_antibodies[cd16_idx, ltq_post_est_sub_df.NK_group==group3],\n",
    "             label=group3, \n",
    "             color= NK_colors[group3],\n",
    "             s=s,\n",
    "             alpha=0.3)\n",
    "\n",
    "\n",
    "main_ax.scatter(x=trnfs_antibodies[cd56_idx, ltq_post_est_sub_df.NK_group==group1],\n",
    "            y=trnfs_antibodies[cd16_idx, ltq_post_est_sub_df.NK_group==group1],\n",
    "             label=group1, \n",
    "             color= NK_colors[group1],\n",
    "             s=s,\n",
    "             alpha=0.3)\n",
    "\n",
    "\n",
    "main_ax.scatter(x=trnfs_antibodies[cd56_idx, ltq_post_est_sub_df.NK_group==group2],\n",
    "            y=trnfs_antibodies[cd16_idx, ltq_post_est_sub_df.NK_group==group2],\n",
    "             label=group2, \n",
    "             color= NK_colors[group2],\n",
    "             s=s,\n",
    "             alpha=0.4)\n",
    "main_ax.set_xlabel(\"CD56\")\n",
    "main_ax.set_ylabel(\"CD16\")\n",
    "\n",
    "# Top histogram (X-axis)\n",
    "x_hist = fig.add_subplot(grid[0, :-1], sharex=main_ax)\n",
    "\n",
    "\n",
    "x_hist.hist(trnfs_antibodies[cd56_idx, ltq_post_est_sub_df.NK_group==group1],\n",
    "         bins=20, label=group1, \n",
    "             color= NK_colors[group1],\n",
    "             alpha=0.5,\n",
    "             density=True)\n",
    "\n",
    "x_hist.hist(trnfs_antibodies[cd56_idx, ltq_post_est_sub_df.NK_group==group2],\n",
    "         bins=20, label=group2, \n",
    "             color= NK_colors[group2],\n",
    "             alpha=0.5,\n",
    "             density=True)\n",
    "\n",
    "x_hist.hist(trnfs_antibodies[cd56_idx, ltq_post_est_sub_df.NK_group==group3],\n",
    "         bins=20, label=group3, \n",
    "             color= NK_colors[group3],\n",
    "             alpha=0.6,\n",
    "             density=True)\n",
    "\n",
    "x_hist.axis(\"off\")  \n",
    "\n",
    "\n",
    "# Side histogram (Y-axis)\n",
    "y_hist = fig.add_subplot(grid[1:, -1], sharey=main_ax)\n",
    "\n",
    "\n",
    "y_hist.hist(trnfs_antibodies[cd16_idx, ltq_post_est_sub_df.NK_group==group1],\n",
    "         bins=20, label=group1, \n",
    "             color= NK_colors[group1],\n",
    "             alpha=0.5,orientation=\"horizontal\",\n",
    "             density=True)\n",
    "\n",
    "y_hist.hist(trnfs_antibodies[cd16_idx, ltq_post_est_sub_df.NK_group==group2],\n",
    "         bins=20, label=group2, \n",
    "             color= NK_colors[group2],\n",
    "             alpha=0.5,orientation=\"horizontal\",\n",
    "             density=True)\n",
    "\n",
    "\n",
    "y_hist.hist(trnfs_antibodies[cd16_idx, ltq_post_est_sub_df.NK_group==group3],\n",
    "         bins=20, label=group3, \n",
    "             color= NK_colors[group3],\n",
    "             alpha=0.6,orientation=\"horizontal\",\n",
    "             density=True)\n",
    "y_hist.axis(\"off\")  \n",
    "\n",
    "plt.subplots_adjust(left=0.1, right=0.9, top=0.9, bottom=0.1)\n",
    "\n",
    "plt.savefig('{}/antibody_expr-scatter-stepfilled-antibody_CD56_CD16-with_hist.svg'.format(save_dir), bbox_inches='tight')\n",
    "plt.savefig('{}/antibody_expr-scatter-stepfilled-antibody_CD56_CD16-with_hist.png'.format(save_dir), bbox_inches='tight')\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "039b33ad-8bd3-4b92-84bd-8d2b2b806688",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_processed[group] = ltq_post_est_sub_df[group]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2fb6ab7-b272-4208-b09b-118b370f651d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(2,2, figsize=(6,4), dpi=300)\n",
    "ax = ax.flatten()\n",
    "\n",
    "# top 4 marker genes for NK from coord blood\n",
    "genes = [\"NKG7\", \n",
    "         \"GNLY\",\n",
    "         \"CMC1\",\n",
    "         \"CST7\"\n",
    "        ]\n",
    "\n",
    "\n",
    "len(genes)\n",
    "for ax_idx, gene in enumerate(genes): \n",
    "    \n",
    "    \n",
    "    label=\"other\"\n",
    "    subset = df_processed[group] ==label\n",
    "    x = df_processed[subset][gene]\n",
    "    ax[ax_idx].hist(x,\n",
    "             bins=20, label=label, \n",
    "                 color= NK_colors[label],\n",
    "                 alpha=0.5,\n",
    "                 density=True)\n",
    "    \n",
    "    label=\"NK_lymphoid\"\n",
    "    subset = df_processed[group] ==label\n",
    "    x = df_processed[subset][gene]\n",
    "    ax[ax_idx].hist(x,\n",
    "             bins=20, label=label, \n",
    "                 color= NK_colors[label],\n",
    "                 alpha=0.5,\n",
    "                    histtype=\"stepfilled\",\n",
    "                 density=True)\n",
    "    \n",
    "    label=\"NK_myeloid\"\n",
    "    subset = df_processed[group] ==label\n",
    "    x = df_processed[subset][gene]\n",
    "    ax[ax_idx].hist(x,\n",
    "             bins=10, label=label, \n",
    "                 color= NK_colors[label],\n",
    "                 alpha=0.5,\n",
    "                    histtype=\"stepfilled\",\n",
    "                 density=True)\n",
    "    \n",
    "    \n",
    "    ax[ax_idx].set_title(gene)\n",
    "fig.text(0.5, 0.0, \"log transcription quotients\", ha='center', va='center')\n",
    "fig.text(0, 0.5, 'density', ha='center', va='center', rotation='vertical')\n",
    "    \n",
    "plt.tight_layout()\n",
    "\n",
    "gene_str = \"_\".join(genes)\n",
    "plt.savefig('{}/marker_gene_expr-histogram-stepfilled-genes_{}.svg'.format(save_dir, gene_str), bbox_inches='tight')\n",
    "plt.savefig('{}/marker_gene_expr-histogram-stepfilled-gene_{}.png'.format(save_dir, gene_str), bbox_inches='tight')\n",
    "    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f8d6d17-75c0-43aa-9e0c-c14aeddc6458",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, ax = plt.subplots(2,2, figsize=(6,4), dpi=300)\n",
    "ax = ax.flatten()\n",
    "\n",
    "genes = [\n",
    "    \"FTL\",\n",
    "    \"CST3\", \n",
    "    \"CTSW\",\n",
    "    \"TRBC2\"\n",
    "]\n",
    "len(genes)\n",
    "for ax_idx, gene in enumerate(genes): \n",
    "    \n",
    "    label=\"NK_lymphoid\"\n",
    "    subset = ltq_post_est_sub_df[group] ==label\n",
    "    x = ltq_post_est_sub_df[subset][gene]\n",
    "    ax[ax_idx].hist(x,\n",
    "             bins=20, label=label, \n",
    "                 color= NK_colors[label],\n",
    "                 alpha=0.5,\n",
    "                    histtype=\"stepfilled\",\n",
    "                 density=True)\n",
    "    \n",
    "    label=\"NK_myeloid\"\n",
    "    subset = ltq_post_est_sub_df[group] ==label\n",
    "    x = ltq_post_est_sub_df[subset][gene]\n",
    "    ax[ax_idx].hist(x,\n",
    "             bins=10, label=label, \n",
    "                 color= NK_colors[label],\n",
    "                 alpha=0.5,\n",
    "                    histtype=\"stepfilled\",\n",
    "                 density=True)\n",
    "    \n",
    "    \n",
    "    ax[ax_idx].set_title(gene)\n",
    "fig.text(0.5, 0.0, \"log transcription quotients\", ha='center', va='center')\n",
    "fig.text(0, 0.5, 'density', ha='center', va='center', rotation='vertical')\n",
    "    \n",
    "plt.tight_layout()\n",
    "\n",
    "gene_str = \"_\".join(genes)\n",
    "plt.savefig('{}/marker_gene_expr-histogram-stepfilled-genes_{}.svg'.format(save_dir, gene_str), bbox_inches='tight')\n",
    "plt.savefig('{}/marker_gene_expr-histogram-stepfilled-gene_{}.png'.format(save_dir, gene_str), bbox_inches='tight')\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5689d73-60ba-4b98-aac1-b20451b88d28",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50fab462-7f4d-42d5-abb9-64eaa390ad96",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bonsai_v3_cloned",
   "language": "python",
   "name": "bonsai_v3_cloned"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
